### PR TITLE
Add Python 3.12 to automated test matrix

### LINKED ISSUE #457: Test against Python 3.12
Currently we test 3.8 to 3.11, but not 3.12.

### PR DESCRIPTION
**Pull request by chuckwondo**
Fixes #457

----
📚 Documentation preview 📚: https://earthaccess--571.org.readthedocs.build/en/571/

### CODE REVIEW

File: .github/workflows/integration-test.yml [hunk 1]
```diff
@@ -35,17 +39,17 @@ jobs:
         id: full-python-version
         run: echo ::set-output name=version::$(python -c "import sys; print('-'.join(str(v) for v in sys.version_info))")
       - name: Install poetry
-        run: |
-          curl -sSL https://install.python-poetry.org | python3 -
-          echo "$HOME/.poetry/bin" >> $GITHUB_PATH
+        uses: abatilo/actions-poetry@v3
```
Comment (mfisher87) [inline .github/workflows/integration-test.yml:42]: :100:

File: .github/workflows/integration-test.yml [hunk 2]
```diff
@@ -59,4 +63,6 @@ jobs:
           EARTHACCESS_TEST_PASSWORD: ${{ secrets.EDL_PASSWORD }}
         run: poetry run bash scripts/integration-test.sh
       - name: Upload coverage
+        # Don't upload coverage when using the `act` tool to run the workflow locally
+        if: ${{ !env.ACT }}
```
Comment (mfisher87) [inline .github/workflows/integration-test.yml:67]: TIL!

File: pyproject.toml [hunk 3]
```diff
@@ -8,7 +8,8 @@ maintainers = [
     "Luis Lopez <betolin@gmail.com>",
     "Joseph H. Kennedy <jhkennedy@alaska.edu>",
     "James Bourbeau <james@coiled.io>",
-    "Matt Fisher <mfisher87@gmail.com>"
+    "Matt Fisher <mfisher87@gmail.com>",
+    "Chuck Daniels <chuck@developmentseed.org>"
```
Comment (mfisher87) [inline pyproject.toml:12]: Good catch :) Forgot there's more to do than just give you permissions!

**Review by mfisher87 — DISMISSED** [review]
Looks great as usual ;)

**Comment by mfisher87** [comment]
Not sure about the build failure.

**Review by mfisher87 — COMMENTED** [review]
Too bad we needed to set up those messy pins :(

I'm especially curious about pyproj, as the included proj database can be materially different in different versions. Would we be able to avoid these "dynamic" pins if we drop support for Python 3.8?

**Comment by chuckwondo** [comment]
> Too bad we needed to set up those messy pins :(
> 
> I'm especially curious about pyproj, as the included proj database can be materially different in different versions. Would we be able to avoid these "dynamic" pins if we drop support for Python 3.8?

Yes, I believe dropping support for Python 3.8 would do the trick, but do we want to drop support prior to its EOL, which is Oct 2024 (see https://devguide.python.org/versions/)?

**Comment by mfisher87** [comment]
Since members of our ecosystem (Numpy + Pyproj) are doing it already, I think it makes sense for us to follow suit. Numpy dropped support for 3.8 [over a year ago](https://numpy.org/neps/nep-0029-deprecation_policy.html). Here's their policy:

>All minor versions of Python released 42 months prior, and at minimum the two latest minor versions are supported. Python support is only dropped in a major/minor version, and never on a patch release.

**Comment by chuckwondo** [comment]
Should I set min versions of numpy and pyproj to be the min versions that support Python 3.12? That is, do we want users using Python 3.9-3.11 to also get the min versions of numpy and pyproj that only support >=3.12, or is that too restrictive for dependency resolution?

More specifically, do we want these (regardless of Python version, user gets only what works for Python >=3.12):

```toml
numpy = { version = ">=1.26.0", optional = true }
pyproj = ">=3.6.1"
```

or these (looser for Python <3.12)?

```toml
numpy = { version = ">=1.26.0", optional = true, python = ">=3.12" }
pyproj = [
  { version = ">=3.5.0", python = "<3.12" },
  { version = ">=3.6.1", python = ">=3.12" }
]
```

cc: @betolink, @jhkennedy

**Comment by chuckwondo** [comment]
I've pinned numpy (optional) and pyproj to their versions that are compatible with Python 3.12. This means that those using Python <3.12 _also_ get those 3.12-compatible versions. Do we want to loosen the pin to allow for earlier versions of numpy and pyproj for those using Python <3.12?

**Comment by mfisher87** [comment]
> Do we want to loosen the pin to allow for earlier versions of numpy and pyproj for those using Python <3.12?

I think so. I think we should make our pins as loose as we can given our dependency policy (for now, safe to assume SPEC0, I think?) to maximize compatibility with users' environments.

We should probably include a pinning policy in our dependency policy :)

**Comment by chuckwondo** [comment]
> > Do we want to loosen the pin to allow for earlier versions of numpy and pyproj for those using Python <3.12?
> 
> I think so. I think we should make our pins as loose as we can given our dependency policy (for now, safe to assume SPEC0, I think?) to maximize compatibility with users' environments.

Thanks for confirming. I suspected as much. I'll adjust accordingly.

**Comment by chuckwondo** [comment]
@mfisher87, I loosened the numpy and pyproj versions for Python <3.12. I believe this should be good to go now.

I set pyproj to its previous pin, but added a new min version for Python >=3.12. For numpy, there was no existing pin, so I added a min version for Python >=3.12 (the version that numpy added support for 3.12), and a min version 2 versions earlier for Python <3.12.

**Comment by chuckwondo** [comment]
@mfisher87, as you requested, so this appears at the top of your queue :-)

**Review by mfisher87 — APPROVED** [review]
LGTM, sorry for the delay. I appreciated the repeated reminder!

**Pull request merged**